File: app/src/main/java/com/sktpj/td2048/GameEngine.kt
package com.sktpj.td2048

import kotlin.math.hypot
import kotlin.math.max
import kotlin.random.Random

class GameEngine(
    private val random: Random = Random.Default,
    initialFormation: List<CharacterDefinition> = StarterRoster.characters,
) {
    companion object {
        private const val INITIAL_ENEMY_HP = 24f
        private const val ENEMIES_PER_WAVE = 7
        private const val ENEMY_SPAWN_SECONDS = 1.30f
        private const val PROJECTILE_SPEED = 1.30f
        private const val BOSS_EVERY_WAVES = 5
        private const val BOSS_WARNING_SECONDS = 5f
        private const val BOSS_SPEED_RATIO = 0.55f
        private const val BOSS_HP_RATIO = 12f
        private const val SLOW_SECONDS = 1.6f
        private const val SLOW_SPEED_RATIO = 0.65f
        private const val HP_DAMAGE_FLASH_SECONDS = 1.2f
        private const val MAX_LOG_ENTRIES = 6

        fun nodePosition(cellIndex: Int): Pair<Float, Float> {
            val row = GameRules.rowOf(cellIndex)
            val col = GameRules.colOf(cellIndex)
            return Pair((col + 0.5f) / 4f, 0.14f + row * 0.205f)
        }

        fun enemyX(enemy: Enemy): Float =
            if (enemy.enemyType == EnemyType.BOSS) 0.5f else (enemy.lane + 0.5f) / 4f
    }

    private var formation = initialFormation.toList()
    private var enemyId = 1
    private var projectileId = 1
    private var spawnTimer = 0f
    private var enemiesSpawnedThisWave = 0
    private var lastBossWave = 0
    private var pendingBossHand: HandType? = null
    private var state = newGameState()

    init {
        require(formation.size == GameRules.CELL_COUNT)
    }

    fun snapshot(): GameSnapshot = state

    fun reset(): GameSnapshot {
        enemyId = 1
        projectileId = 1
        spawnTimer = 0f
        enemiesSpawnedThisWave = 0
        lastBossWave = 0
        pendingBossHand = null
        state = newGameState()
        return state
    }

    fun setFormation(newFormation: List<CharacterDefinition>): GameSnapshot {
        require(newFormation.size == GameRules.CELL_COUNT)
        require(newFormation.map { it.characterId }.distinct().size == GameRules.CELL_COUNT)
        formation = newFormation.toList()
        return reset()
    }

    fun move(direction: Direction): GameSnapshot {
        if (state.gameOverReason != null) return state
        val result = GameRules.moveWithoutSpawn(state.board, direction)
        if (!result.moved) return state
        val nextBoard = GameRules.spawnRandomTile(result.board, random)
        val gameOverReason = if (GameRules.canMove(nextBoard)) null else GameOverReason.BOARD_STUCK
        val nextLog = if (gameOverReason == GameOverReason.BOARD_STUCK) {
            appendLog(state.eventLog, state.elapsedSeconds, "2048盤面が詰まりました", BattleLogTone.WARNING)
        } else {
            state.eventLog
        }
        state = state.copy(
            board = nextBoard,
            score = state.score + result.createdValues.sum(),
            mergeBurst = result.createdValues.sum(),
            gameOverReason = gameOverReason,
            eventLog = nextLog,
        )
        return state
    }

    fun clearMergeBurst(): GameSnapshot {
        if (state.mergeBurst == 0) return state
        state = state.copy(mergeBurst = 0)
        return state
    }

    fun tick(deltaSeconds: Float): GameSnapshot {
        if (state.gameOverReason != null) return state
        val delta = deltaSeconds.coerceIn(0f, 0.05f)
        val elapsedSeconds = state.elapsedSeconds + delta

        var wave = state.wave
        var score = state.score
        var currentHp = state.currentHp
        var bossWarning = state.bossWarning
        var eventLog = state.eventLog
        var hpDamageFlash = state.hpDamageFlash?.let { flash ->
            val remaining = flash.remainingSeconds - delta
            if (remaining > 0f) flash.copy(remainingSeconds = remaining) else null
        }
        var enemies = state.enemies.map { enemy ->
            val remainingSlow = (enemy.slowRemainingSeconds - delta).coerceAtLeast(0f)
            val speedMultiplier = if (enemy.slowRemainingSeconds > 0f) SLOW_SPEED_RATIO else 1f
            enemy.copy(
                progress = enemy.progress + enemy.speed * speedMultiplier * delta,
                slowRemainingSeconds = remainingSlow,
            )
        }

        val leaked = enemies.filter { it.progress >= 1f }
        if (leaked.isNotEmpty()) {
            val leakDamage = leaked.sumOf { it.hp.toInt().coerceAtLeast(1) }
            currentHp = (currentHp - leakDamage).coerceAtLeast(0)
            hpDamageFlash = HpDamageFlash(leakDamage, HP_DAMAGE_FLASH_SECONDS)
            eventLog = appendLog(
                eventLog,
                elapsedSeconds,
                "敵突破  HP -$leakDamage",
                BattleLogTone.WARNING,
            )
            val leakedIds = leaked.mapTo(mutableSetOf()) { it.id }
            enemies = enemies.filterNot { it.id in leakedIds }
        }

        if (currentHp <= 0) {
            eventLog = appendLog(eventLog, elapsedSeconds, "総HPが0になりました", BattleLogTone.WARNING)
            state = state.copy(
                currentHp = 0,
                enemies = enemies,
                gameOverReason = GameOverReason.HP_ZERO,
                elapsedSeconds = elapsedSeconds,
                eventLog = eventLog,
                hpDamageFlash = hpDamageFlash,
            )
            return state
        }

        if (bossWarning != null) {
            val nextRemaining = bossWarning.remainingSeconds - delta
            if (nextRemaining <= 0f) {
                val bossHand = pendingBossHand ?: bossWarning.handType
                enemies = enemies + createBoss(enemyId++, wave, bossHand)
                eventLog = appendLog(
                    eventLog,
                    elapsedSeconds,
                    "BOSS出現  ${handLogLabel(bossHand)}",
                    BattleLogTone.WARNING,
                )
                lastBossWave = wave
                pendingBossHand = null
                bossWarning = null
            } else {
                bossWarning = bossWarning.copy(remainingSeconds = nextRemaining)
            }
        }

        spawnTimer += delta
        val spawnInterval = max(0.48f, ENEMY_SPAWN_SECONDS - (wave - 1) * 0.045f)
        if (spawnTimer >= spawnInterval) {
            spawnTimer = 0f
            enemies = enemies + createNormalEnemy(enemyId++, wave)
            enemiesSpawnedThisWave += 1
            if (enemiesSpawnedThisWave >= ENEMIES_PER_WAVE) {
                enemiesSpawnedThisWave = 0
                wave += 1
                eventLog = appendLog(eventLog, elapsedSeconds, "WAVE $wave", BattleLogTone.INFO)
                if (wave % BOSS_EVERY_WAVES == 0 && lastBossWave != wave) {
                    val hand = randomHand()
                    pendingBossHand = hand
                    bossWarning = BossWarning(BOSS_WARNING_SECONDS, hand)
                    eventLog = appendLog(
                        eventLog,
                        elapsedSeconds,
                        "BOSS WARNING  ${handLogLabel(hand)}",
                        BattleLogTone.WARNING,
                    )
                }
            }
        }

        val cooldowns = state.cooldowns.toMutableList()
        for (index in cooldowns.indices) cooldowns[index] = (cooldowns[index] - delta).coerceAtLeast(0f)

        var projectiles = state.projectiles
        for (cellIndex in 0 until GameRules.CELL_COUNT) {
            val tileValue = state.board[cellIndex]
            if (tileValue <= 0 || cooldowns[cellIndex] > 0f) continue
            val target = TargetingPolicy.selectTarget(cellIndex, enemies) ?: continue
            val character = formation[cellIndex]
            val damage = DamageCalculator.damage(tileValue, character, GameRules.rowOf(cellIndex), target)
            val (sourceX, sourceY) = nodePosition(cellIndex)
            projectiles = projectiles + Projectile(
                id = projectileId++,
                sourceCellIndex = cellIndex,
                sourceCharacterId = character.characterId,
                targetEnemyId = target.id,
                damage = damage,
                x = sourceX,
                y = sourceY,
                speed = PROJECTILE_SPEED,
                handType = character.handType,
                onHitAbility = character.ability,
            )
            cooldowns[cellIndex] = character.attackIntervalMs / 1000f
        }

        val hits = mutableListOf<Projectile>()
        val moving = mutableListOf<Projectile>()
        for (projectile in projectiles) {
            val existingTarget = enemies.firstOrNull { it.id == projectile.targetEnemyId }
            val target = existingTarget ?: TargetingPolicy.selectTarget(projectile.sourceCellIndex, enemies)
            if (target == null) continue

            val targetX = enemyX(target)
            val targetY = target.progress
            val dx = targetX - projectile.x
            val dy = targetY - projectile.y
            val distance = hypot(dx.toDouble(), dy.toDouble()).toFloat().coerceAtLeast(0.0001f)
            val moveDistance = projectile.speed * delta
            val hitRadius = if (target.enemyType == EnemyType.BOSS) 0.065f else 0.035f
            if (distance <= moveDistance + hitRadius) {
                hits += projectile.copy(targetEnemyId = target.id)
            } else {
                moving += projectile.copy(
                    targetEnemyId = target.id,
                    x = projectile.x + (dx / distance) * moveDistance,
                    y = projectile.y + (dy / distance) * moveDistance,
                )
            }
        }

        if (hits.isNotEmpty()) {
            val hitByEnemy = hits.groupBy { it.targetEnemyId }
            val survivors = mutableListOf<Enemy>()
            for (enemy in enemies) {
                val enemyHits = hitByEnemy[enemy.id].orEmpty()
                val damage = enemyHits.sumOf { it.damage }
                val nextHp = enemy.hp - damage
                if (nextHp <= 0f) {
                    score += enemy.maxHp.toInt()
                    if (enemy.enemyType == EnemyType.BOSS) {
                        eventLog = appendLog(
                            eventLog,
                            elapsedSeconds,
                            "BOSS撃破  SCORE +${enemy.maxHp.toInt()}",
                            BattleLogTone.GOOD,
                        )
                    }
                } else {
                    val slowed = enemyHits.any { it.onHitAbility == CharacterAbility.SLOW }
                    survivors += enemy.copy(
                        hp = nextHp,
                        slowRemainingSeconds = if (slowed) max(enemy.slowRemainingSeconds, SLOW_SECONDS) else enemy.slowRemainingSeconds,
                    )
                }
            }
            enemies = survivors
        }

        state = state.copy(
            formation = formation,
            score = score,
            currentHp = currentHp,
            wave = wave,
            enemies = enemies,
            projectiles = moving,
            cooldowns = cooldowns,
            bossWarning = bossWarning,
            elapsedSeconds = elapsedSeconds,
            eventLog = eventLog,
            hpDamageFlash = hpDamageFlash,
        )
        return state
    }

    private fun newGameState(): GameSnapshot {
        val maxHp = FormationRules.maxHp(formation)
        return GameSnapshot(
            board = GameRules.initialBoard(random),
            formation = formation,
            score = 0,
            currentHp = maxHp,
            maxHp = maxHp,
            wave = 1,
            enemies = emptyList(),
            projectiles = emptyList(),
            cooldowns = List(GameRules.CELL_COUNT) { 0f },
            bossWarning = null,
            gameOverReason = null,
            mergeBurst = 0,
            elapsedSeconds = 0f,
            eventLog = listOf(BattleLogEntry(0f, "防衛開始  WAVE 1", BattleLogTone.INFO)),
            hpDamageFlash = null,
        )
    }

    private fun createNormalEnemy(id: Int, wave: Int): Enemy {
        val lane = random.nextInt(GameRules.GRID_SIZE)
        val maxHp = normalEnemyHp(wave) + random.nextInt(0, 9)
        return Enemy(
            id = id,
            enemyType = EnemyType.NORMAL,
            lane = lane,
            progress = -0.05f,
            speed = 0.075f + random.nextFloat() * 0.035f + wave * 0.003f,
            hp = maxHp,
            maxHp = maxHp,
            handType = randomHand(),
        )
    }

    private fun createBoss(id: Int, wave: Int, handType: HandType): Enemy {
        val hp = normalEnemyHp(wave) * BOSS_HP_RATIO
        return Enemy(
            id = id,
            enemyType = EnemyType.BOSS,
            lane = TargetingPolicy.BOSS_LANE,
            progress = -0.08f,
            speed = (0.075f + wave * 0.003f) * BOSS_SPEED_RATIO,
            hp = hp,
            maxHp = hp,
            handType = handType,
        )
    }

    private fun normalEnemyHp(wave: Int): Float = INITIAL_ENEMY_HP + wave * 7f

    private fun randomHand(): HandType = HandType.entries[random.nextInt(HandType.entries.size)]

    private fun appendLog(
        current: List<BattleLogEntry>,
        timestampSeconds: Float,
        message: String,
        tone: BattleLogTone,
    ): List<BattleLogEntry> = (current + BattleLogEntry(timestampSeconds, message, tone)).takeLast(MAX_LOG_ENTRIES)

    private fun handLogLabel(handType: HandType): String = when (handType) {
        HandType.ROCK -> "グー"
        HandType.SCISSORS -> "チョキ"
        HandType.PAPER -> "パー"
    }
}

File: docs/src/game_engine.js
import { GRID_SIZE, canMove, initialBoard, moveWithoutSpawn, spawnRandomTile } from "./game_rules.js";
import { WeaponType, canAttack, columnLevel, columnPower, fireIntervalSeconds, projectileSpeed, selectTarget, weaponType } from "./column_combat_rules.js";

const MAX_HP = 1931;
const INITIAL_ENEMY_HP = 24;
const ENEMIES_PER_WAVE = 7;
const ENEMY_SPAWN_SECONDS = 1.30;
const BOSS_EVERY_WAVES = 5;
const BOSS_WARNING_SECONDS = 5;
const BOSS_SPEED_RATIO = 0.55;
const BOSS_HP_RATIO = 12;
const SAVE_SCHEMA = 1;

export class GameEngine {
  constructor(random = Math.random) {
    this.random = random;
    this.reset();
  }

  reset() {
    this.enemyId = 1;
    this.projectileId = 1;
    this.spawnTimer = 0;
    this.enemiesSpawnedThisWave = 0;
    this.lastBossWave = 0;
    this.pendingBoss = false;
    this.state = {
      schema: SAVE_SCHEMA,
      board: initialBoard(this.random),
      score: 0,
      currentHp: MAX_HP,
      maxHp: MAX_HP,
      wave: 1,
      enemies: [],
      projectiles: [],
      cooldowns: [0, 0, 0, 0],
      bossWarning: null,
      gameOverReason: null,
      mergeBurst: 0,
      elapsedSeconds: 0,
    };
    return this.snapshot();
  }

  snapshot() {
    return structuredClone(this.state);
  }

  restore(saved) {
    if (!saved || typeof saved !== "object") return false;
    const state = saved.state ?? saved;
    if (!Array.isArray(state.board) || state.board.length !== 16) return false;
    if (!Array.isArray(state.enemies) || !Array.isArray(state.projectiles)) return false;
    this.state = {
      schema: SAVE_SCHEMA,
      board: state.board.map((v) => Number(v) || 0),
      score: Math.max(0, Number(state.score) || 0),
      currentHp: Math.max(0, Number(state.currentHp) || MAX_HP),
      maxHp: Math.max(1, Number(state.maxHp) || MAX_HP),
      wave: Math.max(1, Number(state.wave) || 1),
      enemies: state.enemies.map((e) => ({ ...e })),
      projectiles: state.projectiles.map((p) => ({ ...p })),
      cooldowns: Array.isArray(state.cooldowns) ? state.cooldowns.slice(0, 4).map((v) => Math.max(0, Number(v) || 0)) : [0,0,0,0],
      bossWarning: state.bossWarning ? { remainingSeconds: Math.max(0, Number(state.bossWarning.remainingSeconds) || 0) } : null,
      gameOverReason: state.gameOverReason === "BOARD_STUCK" || state.gameOverReason === "HP_ZERO" ? state.gameOverReason : null,
      mergeBurst: 0,
      elapsedSeconds: Math.max(0, Number(state.elapsedSeconds) || 0),
    };
    while (this.state.cooldowns.length < 4) this.state.cooldowns.push(0);
    this.enemyId = Math.max(1, ...this.state.enemies.map((e) => Number(e.id) || 0)) + 1;
    this.projectileId = Math.max(1, ...this.state.projectiles.map((p) => Number(p.id) || 0)) + 1;
    this.spawnTimer = Math.max(0, Number(saved.spawnTimer) || 0);
    this.enemiesSpawnedThisWave = Math.max(0, Math.min(ENEMIES_PER_WAVE - 1, Number(saved.enemiesSpawnedThisWave) || 0));
    this.lastBossWave = Math.max(0, Number(saved.lastBossWave) || 0);
    this.pendingBoss = Boolean(saved.pendingBoss ?? this.state.bossWarning);
    return true;
  }

  serialize() {
    return {
      schema: SAVE_SCHEMA,
      state: this.snapshot(),
      spawnTimer: this.spawnTimer,
      enemiesSpawnedThisWave: this.enemiesSpawnedThisWave,
      lastBossWave: this.lastBossWave,
      pendingBoss: this.pendingBoss,
    };
  }

  move(direction) {
    if (this.state.gameOverReason) return { changed: false, scoreChanged: false, gameOver: true };
    const result = moveWithoutSpawn(this.state.board, direction);
    if (!result.moved) return { changed: false, scoreChanged: false, gameOver: false };
    const board = spawnRandomTile(result.board, this.random);
    const gained = result.createdValues.reduce((sum, value) => sum + value, 0);
    this.state.board = board;
    this.state.score += gained;
    this.state.mergeBurst = gained;
    if (!canMove(board)) this.state.gameOverReason = "BOARD_STUCK";
    return { changed: true, scoreChanged: gained > 0, gameOver: Boolean(this.state.gameOverReason) };
  }

  tick(deltaSeconds) {
    if (this.state.gameOverReason) return { scoreChanged: false, waveChanged: false, gameOver: true };
    const delta = Math.max(0, Math.min(0.05, deltaSeconds));
    this.state.elapsedSeconds += delta;
    let scoreChanged = false;
    let waveChanged = false;

    for (const enemy of this.state.enemies) enemy.progress += enemy.speed * delta;
    const leaked = this.state.enemies.filter((enemy) => enemy.progress >= 1);
    if (leaked.length) {
      const leakedIds = new Set(leaked.map((enemy) => enemy.id));
      const damage = leaked.reduce((sum, enemy) => sum + Math.max(1, Math.trunc(enemy.hp)), 0);
      this.state.currentHp = Math.max(0, this.state.currentHp - damage);
      this.state.enemies = this.state.enemies.filter((enemy) => !leakedIds.has(enemy.id));
    }
    if (this.state.currentHp <= 0) {
      this.state.gameOverReason = "HP_ZERO";
      return { scoreChanged, waveChanged, gameOver: true };
    }

    if (this.state.bossWarning) {
      this.state.bossWarning.remainingSeconds -= delta;
      if (this.state.bossWarning.remainingSeconds <= 0) {
        this.state.enemies.push(this.createBoss(this.enemyId++, this.state.wave));
        this.lastBossWave = this.state.wave;
        this.pendingBoss = false;
        this.state.bossWarning = null;
      }
    }

    this.spawnTimer += delta;
    const spawnInterval = Math.max(0.48, ENEMY_SPAWN_SECONDS - (this.state.wave - 1) * 0.045);
    if (this.spawnTimer >= spawnInterval) {
      this.spawnTimer = 0;
      this.state.enemies.push(this.createNormalEnemy(this.enemyId++, this.state.wave));
      this.enemiesSpawnedThisWave += 1;
      if (this.enemiesSpawnedThisWave >= ENEMIES_PER_WAVE) {
        this.enemiesSpawnedThisWave = 0;
        this.state.wave += 1;
        waveChanged = true;
        if (this.state.wave % BOSS_EVERY_WAVES === 0 && this.lastBossWave !== this.state.wave) {
          this.pendingBoss = true;
          this.state.bossWarning = { remainingSeconds: BOSS_WARNING_SECONDS };
        }
      }
    }

    this.state.cooldowns = this.state.cooldowns.map((value) => Math.max(0, value - delta));
    for (let column = 0; column < GRID_SIZE; column += 1) {
      const power = columnPower(this.state.board, column);
      if (power <= 0 || this.state.cooldowns[column] > 0) continue;
      const target = selectTarget(column, this.state.enemies);
      if (!target) continue;
      const type = weaponType(columnLevel(this.state.board, column));
      const source = this.turretPosition(column);
      this.state.projectiles.push({
        id: this.projectileId++, sourceColumn: column, targetEnemyId: target.id,
        damage: power, x: source.x, y: source.y, speed: projectileSpeed(type), weaponType: type,
      });
      this.state.cooldowns[column] = fireIntervalSeconds(type);
    }

    const hits = [];
    const moving = [];
    for (const projectile of this.state.projectiles) {
      const existing = this.state.enemies.find((enemy) => enemy.id === projectile.targetEnemyId);
      const target = existing ?? selectTarget(projectile.sourceColumn, this.state.enemies);
      if (!target) continue;
      const tx = this.enemyX(target);
      const ty = target.progress;
      const dx = tx - projectile.x;
      const dy = ty - projectile.y;
      const distance = Math.max(0.0001, Math.hypot(dx, dy));
      const moveDistance = projectile.speed * delta;
      const hitRadius = target.enemyType === "BOSS" ? 0.065 : 0.035;
      if (distance <= moveDistance + hitRadius) {
        hits.push({ ...projectile, targetEnemyId: target.id });
      } else {
        moving.push({ ...projectile, targetEnemyId: target.id, x: projectile.x + dx / distance * moveDistance, y: projectile.y + dy / distance * moveDistance });
      }
    }
    this.state.projectiles = moving;

    if (hits.length) {
      const damageByEnemyId = new Map();
      const addDamage = (id, damage) => damageByEnemyId.set(id, (damageByEnemyId.get(id) ?? 0) + damage);
      for (const projectile of hits) {
        const target = this.state.enemies.find((enemy) => enemy.id === projectile.targetEnemyId);
        if (!target) continue;
        if ([WeaponType.NORMAL, WeaponType.RAPID, WeaponType.MACHINE_GUN].includes(projectile.weaponType)) {
          addDamage(target.id, projectile.damage);
        } else if (projectile.weaponType === WeaponType.PIERCING) {
          addDamage(target.id, projectile.damage);
          this.state.enemies
            .filter((enemy) => enemy.id !== target.id && canAttack(projectile.sourceColumn, enemy))
            .sort((a, b) => b.progress - a.progress).slice(0, 2)
            .forEach((enemy) => addDamage(enemy.id, Math.max(1, Math.trunc(projectile.damage * 0.70))));
        } else if (projectile.weaponType === WeaponType.EXPLOSIVE) {
          addDamage(target.id, projectile.damage);
          this.state.enemies
            .filter((enemy) => enemy.id !== target.id && canAttack(projectile.sourceColumn, enemy) && Math.abs(enemy.progress - target.progress) <= 0.14)
            .forEach((enemy) => addDamage(enemy.id, Math.max(1, Math.trunc(projectile.damage * 0.60))));
        } else if (projectile.weaponType === WeaponType.LASER) {
          this.state.enemies.filter((enemy) => canAttack(projectile.sourceColumn, enemy)).forEach((enemy) => addDamage(enemy.id, projectile.damage));
        }
      }
      const survivors = [];
      for (const enemy of this.state.enemies) {
        const damage = damageByEnemyId.get(enemy.id) ?? 0;
        if (damage <= 0) { survivors.push(enemy); continue; }
        const nextHp = enemy.hp - damage;
        if (nextHp <= 0) {
          this.state.score += Math.trunc(enemy.maxHp);
          scoreChanged = true;
        } else survivors.push({ ...enemy, hp: nextHp });
      }
      this.state.enemies = survivors;
    }

    return { scoreChanged, waveChanged, gameOver: Boolean(this.state.gameOverReason) };
  }

  createNormalEnemy(id, wave) {
    const lane = Math.floor(this.random() * GRID_SIZE);
    const maxHp = INITIAL_ENEMY_HP + wave * 7 + Math.floor(this.random() * 9);
    return { id, enemyType: "NORMAL", lane, progress: -0.05, speed: 0.075 + this.random() * 0.035 + wave * 0.003, hp: maxHp, maxHp };
  }

  createBoss(id, wave) {
    const hp = (INITIAL_ENEMY_HP + wave * 7) * BOSS_HP_RATIO;
    return { id, enemyType: "BOSS", lane: -1, progress: -0.08, speed: (0.075 + wave * 0.003) * BOSS_SPEED_RATIO, hp, maxHp: hp };
  }

  turretPosition(column) { return { x: (column + 0.5) / GRID_SIZE, y: 0.955 }; }
  enemyX(enemy) { return enemy.enemyType === "BOSS" ? 0.5 : (enemy.lane + 0.5) / GRID_SIZE; }
}

export const CURRENT_MAX_HP = MAX_HP;
